src/org/recompile/mobile/WavImaAdpcmDecoder.java

```diff
@@ -0,0 +1,370 @@
+		short bitsPerSample = (short) readInputStreamInt16(input);
+		String data = readInputStreamASCII(input, 4);
+		int sampleDataLength = readInputStreamInt32(input);
+
+		/* Those are only meant for debugging. */
+		/*
+		System.out.println("WAV HEADER_START");
+
+		System.out.println(riff);
+		System.out.println("FileSize:" + dataSize);
+		System.out.println("Format: " + format);
+
+		System.out.println("---'fmt' header---\n");
+		System.out.println("Header ChunkSize:" + Integer.toString(chunkSize));
+		System.out.println("AudioFormat: " + Integer.toString(audioFormat));
+		System.out.println("AudioChannels:" + Integer.toString(audioChannels));
+		System.out.println("SampleRate:" + Integer.toString(sampleRate));
+		System.out.println("BytesPerSec:" + Integer.toString(bytesPerSec));
+		System.out.println("FrameSize:" + Integer.toString(frameSize));
+		System.out.println("BitsPerSample:" + Integer.toHexString(bitsPerSample));
+
+		System.out.println("\n---'data' header---\n");
+		System.out.println("SampleData Length:" + Integer.toString(sampleDataLength));
+
+		System.out.println("WAV HEADER_END\n\n\n");
+		*/
+		
+		/* 
+		 * We need the audio format to check if it's ADPCM or PCM, and the file's 
+		 * dataSize + SampleRate + audioChannels to decode ADPCM and build the new header. 
+		 */
+		return new int[] {audioFormat, dataSize, sampleRate, audioChannels};
+	}
+
+	/* Read a 16-bit little-endian unsigned integer from input.*/
+	public static int readInputStreamInt16(InputStream input) throws IOException 
+	{ return ( input.read() & 0xFF ) | ( ( input.read() & 0xFF ) << 8 ); }
+
+	/* Read a 32-bit little-endian signed integer from input.*/
+	public static int readInputStreamInt32(InputStream input) throws IOException 
+	{
+		return ( input.read() & 0xFF ) | ( ( input.read() & 0xFF ) << 8 )
+			| ( ( input.read() & 0xFF ) << 16 ) | ( ( input.read() & 0xFF ) << 24 );
+	}
+
+	/* Return a String containing 'n' Characters of ASCII/ISO-8859-1 text from input. */
+	public static String readInputStreamASCII(InputStream input, int nChars) throws IOException 
+	{
+		byte[] chars = new byte[nChars];
+		readInputStreamData(input, chars, 0, nChars);
+		return new String(chars, "ISO-8859-1");
+	}
+
+	/* Read 'n' Bytes from the InputStream starting from the specified offset into the output array. */
+	public static void readInputStreamData(InputStream input, byte[] output, int offset, int nBytes) throws IOException 
+	{
+		int end = offset + nBytes;
+		while(offset < end) 
+		{
+			int read = input.read(output, offset, end - offset);
+			if(read < 0) throw new java.io.EOFException();
+			offset += read;
+		}
+	}
+
+	/* 
+	 * Builds a WAV header that describes the decoded ADPCM file on the first 44 bytes. 
+	 * Data: little-endian, 16-bit, signed, same sample rate and channels as source IMA ADPCM.
+	 */
+	private void buildHeader(byte[] buffer, int numChannels, int sampleRate) 
+	{ 
+		final short bitsPerSample = 16;   /* 16-bit PCM */
```

Comment:
Why is everything final?

Reply:
This method only builds the wav header one time per call, so those values don't really change after their first declaration and they're never used outside of this method, which is why they're final. 

Would that be a problem? Currently i'm treating those the same way as a 'const' in C... I'm open to suggestions if there's a better way to handle this task.